base_layer/core/src/covenants/covenant.rs

```diff
@@ -0,0 +1,160 @@
+//  Copyright 2021, The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{io, iter::FromIterator};
+
+use crate::{
+    common::byte_counter::ByteCounter,
+    consensus::{ConsensusDecoding, ConsensusEncoding, ConsensusEncodingSized},
+    covenants::{
+        context::CovenantContext,
+        decoder::{CovenantDecodeError, CovenantTokenDecoder},
+        encoder::CovenantTokenEncoder,
+        error::CovenantError,
+        filters::Filter,
+        output_set::OutputSet,
+        token::{CovenantToken, CovenantTokenCollection},
+    },
+    transactions::transaction::{TransactionInput, TransactionOutput},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct Covenant {
+    tokens: Vec<CovenantToken>,
+}
+
+impl Covenant {
+    pub fn new() -> Self {
+        Self { tokens: Vec::new() }
+    }
+
+    pub fn from_bytes(mut bytes: &[u8]) -> Result<Self, CovenantDecodeError> {
+        if bytes.is_empty() {
+            return Ok(Self::new());
+        }
+        CovenantTokenDecoder::new(&mut bytes).collect()
+    }
+
+    pub(super) fn write_to<W: io::Write>(&self, writer: &mut W) -> Result<usize, io::Error> {
+        CovenantTokenEncoder::new(self.tokens.as_slice()).write_to(writer)
+    }
+
+    pub fn validate(&self) -> Result<(), CovenantError> {
+        // TODO
```

Comment:
is this still todo for this PR? or as part of the integration

Reply:
So it's slightly towards a nice-to-have, because as it is executed if it is not valid, e.g. has invalid arguments for the filter, the block/txn validation will fail anyway - this is more about validating the correctness of the covenant which isn't so NB for now. Maybe I should remove this funtion.